### ISSUE TITLE
NRI team meeting notes on Nov. 26

### ISSUE THREAD

**Issue by PaulSpence** [PaulSpence/OM2-emulator#5]
Attendees: Micael, Navid, Taimoor, Ryan, Paul, Yue, Edward, Vassilli, Tennessee

**Comment by PaulSpence** [comment]
Notes (by Paul):

NRI can help with technical work and architecture. Science team needs to outline aims. This can be an opportunity to help evolve the working group.

What is the first step?

Taimoor: What is NRI contribution vs working group vs science team? If we have a clearly defined science goal. We can data wrangle the inputs and outputs. We need help with the specifics of machine architecture. Not sure if we have a clear answer. We need to be flexible. The boundaries are not very clear. More a collaboration than a support request.  Perhaps with more regular meetings.

So far we are the only group asking for Edwards help, but that may change.

We should hear from Tennesse and Yue since the are ML experts. They can help with data pipeline using pyearth tools.
 
Tennesse: what are the definitions? Framework, pytorch or pyearth. The model is trained the neural network. Architecture is the design of the neural network. We have to work out if what we are doing is novel. There is probably some pre-existing literature. What are people already doing? Concept is solid to build an emulator, but is it relavent. Start by building an auto-encorder. All the other architectures are footnotes to auto-encorder with pyearth tools. What is a good verification score card for this idea? Questions about who does what. The working group meetings are good place to present lit review and next step ideas. My role is guidance and support on how do use pyearth tools. More intense level of support would have to come from others.

Yue: Suggest science team start with existing models. Replicate a model to start. For example: https://arxiv.org/abs/2509.00015 https://github.com/CMCC-Foundation/MedFormer links maybe broken. Start with there training code to replicate published results as first step. This sorts out many important steps. If we want some processes to be captured than we have to use the architecture that can do that. What is actually essential for this emulator to be scientifically valid? Before release to community, do we want to try different architectures first, or just present a first emulation? Start with something already done and replicate it, then feed in our own data set. Edward can help make the data loader more efficient. Then we can all contribute to deciding what architecture is best. If we can make re-inforcement learning happen, that would be great and novel. Likes the idea of reinforcement learning. 

Taimoor: re literature and relevance aspect. Can take a previous architecture and use it, but sees benefit in creating our own path. Useful to have domestic expertise.  Big gaps now are in architectures that can do future forecasting. 

Tennessee: its not hard to change the architecture. Makes sense to make Australia the focus. Get the inputs and outputs right, then can play with architectures.  What are the science challenges we need the model to solve? Awesome list: 

Edward: Discussion is about project design at this stage. Doesn't think this as the NRI stage. IF there is an ML took that works, NRI can help make the tool more accessible in terms of standardizing inputs and outputs. Further down the line. What is the value of having an Australian focussed emulator? First speak with Tennesse and Yue for ML help.

Ryan: What does the Australian context mean? The ocean models are massively expensive. The object is to make future projections or spinup way less expensive. If we can help make these expts cheaper that will help a lot. We need more ensembles at higher resolution. 

Vassilli: QuickClim - our CMIP5 emulator that does just that https://www.nature.com/articles/s43247-023-01011-0. Emulators that can work with all of CMIP5 data.

Ocean 3D emulator https://link.springer.com/article/10.1007/s00162-024-00719-9 EOF stuff. Kind of a spectral emulator.

A different science area but this might be of interest: https://github.com/hoonerg/Awesome-Diffusion-Models-for-Weather-Forecasting

Ryan/Paul: The motivation is to be able to do 1000s of ensembles of om2-01 simulations.

Micael: clear interest and motivation. Its a good project to have under the NRI umbrella. Goal of NRI is to build the ML working group: joining scientists with NRI team to build the working group. Path forward to have another meeting.  What are the next steps?

Show up at the meetings. Drill down on the outcomes. Simple and easy: go do the autoencorder tutorials! Simplifies the problem. Then we will have a common language to build from. Building the first model is not the hard part. The next step is to get the datasets to start with (inputs surface heat salt flux and outputs integrated heat and salt from om2-01). Next week there is a working group meeting, Taimoor present a bit?

Next steps (no order):
i) Find ocean emulator papers
ii) do encorder tutorial
iii) schedule another emulator meeting
iv) attend the ML WG meeting next week
v) work on the data pipeline (inputs and outputs)

**Comment by rmholmes** [comment]
## **Notes, team meeting 29/1/2026**

Taimoor has a great first pass going which uses PET to bring in the OM2 OHC and Qnet data that Paul created. We went through the notebook and identified issues, questions and next steps. Issues were created and assigned.

Identified issues:
- air-sea heat flux is positive definite.
- The reconstruction is not dealing with land very well.
- Change in PET with the `AlignDataVariableDimensionsToDatasetCoords` function.
- Taimoor had to do the normalization by hand (inefficient) rather than using PET.

Questions we have:
- Why is the predicted air-sea heat flux positive definite? This is not related to land-sea masking.
- How do we put our normalized L1 loss in context? What does it mean? What does reducing it mean? How well is our model working? Does the reconstruction conserve total (ocean-integrated) OHC?
- Do we need to worry about the `Sort` and `AlignDataVariableDimensionsToDatasetCoords` functions in the pipeline definition?

Things to explore and next steps:
- Partial, mask-aware, convolutions (Ryan has code for this already)
- Land-aware loss masking - we only want the ocean to contribute to the loss (again, Ryan has code for this already).
- Play around with the architecture depth, kernel sizes etc.
- Do the normalization with PET rather than by hand.

**Comment by rmholmes** [comment]
**Notes, team meeting 2/4/2026:**
---

- Ryan did some work to:
   1) clean-up the notebook for easier debugging, including separating out the pipeline code into a separate python script.
   2) implement partial convolutions with mask evolution.
   3) implement land-aware loss masking
- The partial convolutions and land-aware masking appear to overall be working. However, there were still some problems with training the model when using the actual land-sea mask (fine with a 0 mask) - Ryan can't remember exactly what the problem was, but at the time thought that it was related to the fact that we are not normalising properly.
- The team discussed pro's/con's of the mask evolution (through the upsampling/downsampling in the decoder-encoder architecture) - in particular the fact that land areas "disappear" under downsampling. This needs more thought.
- The team discussed Taimoor's DP proposal ideas.

Ryan has merged the code updates into the main branch.

**Key next step:** Get proper normalisation working (needed to confirm that partial convolutions are working).
